Kani's assert macro implementation bypasses some compile-time checks

<!--
Please report security concerns to us on the
[AWS vulnerability reporting](http://aws.amazon.com/security/vulnerability-reporting/) page.
-->

<!--
Thank you for filing a bug report! 🐛
Please provide a short summary of the issue, along with the information necessary to replicate.
-->

With PR #797, Kani's implementation of the assert macro's second form that accepts a custom message bypasses several compile-time checks. 

For example, while `rustc` errors out on the following program:
<!--
If the reproducing code is small, please post it here.
Otherwise, please attach the relevant files to this issue.
Alternatively, post a link to the repository and branch that exposes the issue.
-->
```rust
fn main() {
    assert!(1 + 1 == 3, "Bad arithmetic: {}", foo);
}
```
output:
```
error[E0425]: cannot find value `foo` in this scope
 --> no_display.rs:2:47
  |
2 |     assert!(1 + 1 == 3, "Bad arithmetic: {}", foo);
  |                                               ^^^ not found in this scope
```
kani does not complain about the non-existence of the variable `foo`:
```
CBMC version 5.48.0 (cbmc-5.48.0) 64-bit x86_64 linux
Reading GOTO program from file no_display.goto
Generating GOTO Program
Adding CPROVER library (x86_64)
Removal of function pointers and virtual functions
Generic Property Instrumentation
Running with 16 object bits, 48 offset bits (user-specified)
Starting Bounded Model Checking
Runtime Symex: 0.00102581s
size of program expression: 75 steps
simple slicing removed 3 assignments
Generated 3 VCC(s), 1 remaining after simplification
Runtime Postprocess Equation: 7.25e-06s
Passing problem to propositional reduction
converting SSA
Runtime Convert SSA: 0.000213004s
Running propositional reduction
Post-processing
Runtime Post-process: 2.594e-05s
Solving with MiniSAT 2.2.1 with simplifier
100 variables, 33 clauses
SAT checker: instance is SATISFIABLE
Runtime Solver: 9.1981e-05s
Runtime decision procedure: 0.000343287s

** Results:
[assertion.1] Code generation sanity check: Correct CBMC vtable size for StructTag("tag-_4317141983261250885") (MIR type Closure(DefId(2:9147 ~ std[2f3c]::rt::lang_start::{closure#0}), [(), i8, extern "rust-call" fn(()) -> i32, (fn(),)])). Please report failures:
https://github.com/model-checking/kani/issues/new?template=bug_report.md: SUCCESS
[pointer_primitives.1] dead object in OBJECT_SIZE(&temp_0): SUCCESS
/home/ubuntu/examples/no_display.rs function main
[main.assertion.1] line 2 assertion: "Bad arithmetic: {}", foo: FAILURE

** 1 of 3 failed (2 iterations)
VERIFICATION FAILED
```
Other checks that would be skipped by kani's assert macro are that the argument must implement the `Display` trait.

--- a/library/std/src/lib.rs
+++ b/library/std/src/lib.rs
@@ -34,18 +34,28 @@ pub mod process;
 /// assert!(a + b == c, "The sum of {} and {} is {}", a, b, c);
 /// ```
 /// the assert message will be:
-/// "The sum of {} and {} is {}", 1, 1, 2
+/// "The sum of {} and {} is {}", a, b, c
 #[macro_export]
 macro_rules! assert {
     ($cond:expr $(,)?) => {
         kani::assert($cond, concat!("assertion failed: ", stringify!($cond)));
     };
     ($cond:expr, $($arg:tt)+) => {
-        // Note that by stringifying the arguments to the custom message, any
-        // compile-time checks on those arguments (e.g. checking that the symbol
-        // is defined and that it implements the Display trait) are bypassed:
-        // https://github.com/model-checking/kani/issues/803
         kani::assert($cond, concat!(stringify!($($arg)+)));
+        // Process the arguments of the assert inside an unreachable block. This
+        // is to make sure errors in the arguments (e.g. an unknown variable or
+        // an argument that does not implement the Display or Debug traits) are
+        // reported, without creating any overhead on verification performance
+        // that may arise from processing strings involved in the arguments.
+        // Note that this approach is only correct with the "abort" panic
+        // strategy, but is unsound with the "unwind" panic strategy which
+        // requires evaluating the arguments (because they might have side
+        // effects). This is fine until we add support for the "unwind" panic
+        // strategy, which is tracked in
+        // https://github.com/model-checking/kani/issues/692
+        if false {
+            let _ = format_args!($($arg)+);
+        }
     };
 }
 
@@ -100,7 +110,7 @@ macro_rules! debug_assert_ne {
     ($($x:tt)*) => ({ $crate::assert_ne!($($x)*); })
 }
 
-// Override the print macros to skip all the formatting functionality (which
+// Override the print macros to skip all the printing functionality (which
 // is not relevant for verification)
 #[macro_export]
 macro_rules! print {

--- a/tests/expected/assert-arg-error/expected
+++ b/tests/expected/assert-arg-error/expected
@@ -1,0 +1,1 @@
+cannot find value `foo` in this scope

--- a/tests/expected/assert-arg-error/test.rs
+++ b/tests/expected/assert-arg-error/test.rs
@@ -1,0 +1,10 @@
+// Copyright Kani Contributors
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+
+//! This test checks that Kani processes arguments of assert macros and produces
+//! an error for invalid arguments (e.g. unknown variable)
+
+#[kani::proof]
+fn check_invalid_value_error() {
+    assert!(1 + 1 == 2, "An assertion message that references an unknown variable {}", foo);
+}

--- a/tests/kani/Assert/var_in_arg.rs
+++ b/tests/kani/Assert/var_in_arg.rs
@@ -1,0 +1,16 @@
+// Copyright Kani Contributors
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+
+//! This test makes sure Kani does not emit "unused variable" warnings for
+//! variables that are only used in arguments of assert macros
+
+// Promote "unused variable" warnings to an error so that this test fails if
+// Kani's overridden version of the assert macros drops variables used as
+// arguments of those macros
+#![deny(unused_variables)]
+
+#[kani::proof]
+fn check_assert_with_arg() {
+    let s = "foo";
+    assert!(1 + 1 == 2, "An assertion message that refers to a variable {}", s);
+}

